File: module-telegram/src/sessionManager.ts
/**
 * module-telegram/src/sessionManager.ts
 *
 * Gerencia sessões Telegram via MTProto (GramJS) — User API, não Bot.
 * Fluxo idêntico ao WhatsApp:
 *   1. Frontend pede /session → gera código SMS/app
 *   2. Frontend envia /session/confirm com o código → sessão conecta
 *   3. /qr/:sessionId → retorna status + phone number confirmado
 *   4. sendMessage(sessionId, phone, text) → envia como usuário real
 */

import { TelegramClient } from "telegram"
import { StringSession } from "telegram/sessions"
import { Api } from "telegram"
import { redis, supabase, logger } from "@zapflow/shared"
import path from "path"
import fs from "fs"

// ── Armazenamento em memória ────────────────────────────────────────────────
const sessions = new Map<string, TelegramClient>()

// Sessões aguardando confirmação do código SMS
const pendingAuth = new Map<string, {
  client:      TelegramClient
  phoneNumber: string
  phoneHash:   string
  ownerId:     string
}>()

const SESSION_DIR = process.env.TG_SESSION_DIR
  ?? path.join(process.cwd(), ".tg-sessions")

fs.mkdirSync(SESSION_DIR, { recursive: true })

function sessionFilePath(sessionId: string): string {
  return path.join(SESSION_DIR, `${sessionId}.session`)
}

function loadStringSession(sessionId: string): string {
  const fp = sessionFilePath(sessionId)
  if (fs.existsSync(fp)) return fs.readFileSync(fp, "utf8").trim()
  return ""
}

function saveStringSession(sessionId: string, str: string): void {
  fs.writeFileSync(sessionFilePath(sessionId), str, "utf8")
}

// ── Criar cliente GramJS ────────────────────────────────────────────────────
function makeClient(sessionStr = ""): TelegramClient {
  const apiId   = Number(process.env.TELEGRAM_API_ID)
  const apiHash = process.env.TELEGRAM_API_HASH ?? ""

  if (!apiId || !apiHash) {
    throw new Error("TELEGRAM_API_ID e TELEGRAM_API_HASH são obrigatórios no .env")
  }

  return new TelegramClient(
    new StringSession(sessionStr),
    apiId,
    apiHash,
    {
      connectionRetries: 5,
      deviceModel:       "ZapFlow",
      systemVersion:     "Linux",
      appVersion:        "2.0.0",
      langCode:          "pt",
    }
  )
}

// ── Passo 1 — iniciar login (envia código para o celular) ───────────────────
export async function startLogin(
  sessionId: string,
  ownerId:   string,
  phone:     string
): Promise<void> {
  // Se já existe sessão ativa, não faz nada
  if (sessions.has(sessionId)) {
    logger.info({ sessionId }, "Sessão Telegram já ativa")
    return
  }

  const savedStr = loadStringSession(sessionId)
  const client   = makeClient(savedStr)

  await client.connect()

  if (await client.isUserAuthorized()) {
    // Sessão salva em disco ainda é válida — restaura direto
    sessions.set(sessionId, client)
    await setStatus(sessionId, ownerId, phone, "connected")
    logger.info({ sessionId, phone }, "Sessão Telegram restaurada do disco")
    return
  }

  // Solicita o código ao Telegram
  const { phoneCodeHash } = await client.sendCode(
    { apiId: Number(process.env.TELEGRAM_API_ID), apiHash: process.env.TELEGRAM_API_HASH! },
    phone
  )

  pendingAuth.set(sessionId, { client, phoneNumber: phone, phoneHash: phoneCodeHash, ownerId })

  // Salva no Redis para o frontend saber que está aguardando código
  await redis.set(`tg:pending:${sessionId}`, phone, "EX", 300)

  logger.info({ sessionId, phone }, "Código Telegram enviado — aguardando confirmação")
}

// ── Passo 2 — confirmar código recebido no celular/app ──────────────────────
export async function confirmCode(
  sessionId: string,
  code:      string,
  password?: string   // 2FA — se a conta tiver senha de dois fatores
): Promise<void> {
  const pending = pendingAuth.get(sessionId)
  if (!pending) throw new Error("Nenhum login pendente para esta sessão")

  const { client, phoneNumber, phoneHash, ownerId } = pending

  try {
    await client.invoke(new Api.auth.SignIn({
      phoneNumber,
      phoneCodeHash: phoneHash,
      phoneCode:     code,
    }))
  } catch (err: any) {
    // Conta com 2FA — envia senha
    if (err?.message?.includes("SESSION_PASSWORD_NEEDED")) {
      if (!password) throw new Error("Esta conta tem 2FA. Envie a senha no campo 'password'.")
      const { SRPParams } = await client.invoke(new Api.account.GetPassword())
      // @ts-ignore — computeCheck existe no gramjs mas falta tipagem
      const check = await (client as any).computePasswordSRP(SRPParams, password)
      await client.invoke(new Api.auth.CheckPassword({ password: check }))
    } else {
      throw err
    }
  }

  // Persiste a session string em disco (como o WhatsApp persiste .wa-auth)
  const sessionStr = (client.session as StringSession).save()
  saveStringSession(sessionId, sessionStr)

  sessions.set(sessionId, client)
  pendingAuth.delete(sessionId)
  await redis.del(`tg:pending:${sessionId}`)

  await setStatus(sessionId, ownerId, phoneNumber, "connected")
  logger.info({ sessionId, phoneNumber }, "Sessão Telegram conectada com sucesso")
}

// ── Enviar mensagem por número de telefone ──────────────────────────────────
export async function sendMessage(
  sessionId: string,
  to:        string,   // número E.164: 5581994900228
  text:      string
): Promise<boolean> {
  let client = sessions.get(sessionId)

  if (!client) {
    // Tenta restaurar do disco
    const savedStr = loadStringSession(sessionId)
    if (!savedStr) {
      logger.warn({ sessionId }, "Sessão Telegram não encontrada")
      return false
    }

    const { data } = await supabase
      .from("telegram_sessions")
      .select("owner_id,phone_number")
      .eq("session_id", sessionId)
      .maybeSingle()

    if (!data?.owner_id) {
      logger.warn({ sessionId }, "owner_id não encontrado no Supabase")
      return false
    }

    client = makeClient(savedStr)
    await client.connect()

    if (!(await client.isUserAuthorized())) {
      logger.warn({ sessionId }, "Sessão expirada — reconectar necessário")
      await setStatus(sessionId, data.owner_id, data.phone_number, "disconnected")
      sessions.delete(sessionId)
      return false
    }

    sessions.set(sessionId, client)
    logger.info({ sessionId }, "Sessão Telegram restaurada para envio")
  }

  try {
    // GramJS aceita o número com + ou sem — normaliza aqui
    const recipient = to.startsWith("+") ? to : `+${to}`
    await client.sendMessage(recipient, { message: text })
    logger.info({ sessionId, to: maskPhone(to) }, "✓ Telegram MTProto enviado")
    return true
  } catch (err: any) {
    logger.error({ sessionId, to: maskPhone(to), err: err.message }, "✗ Falha Telegram")

    // Sessão revogada remotamente
    if (err?.message?.includes("AUTH_KEY_UNREGISTERED") ||
        err?.message?.includes("SESSION_REVOKED")) {
      sessions.delete(sessionId)
      const { data } = await supabase
        .from("telegram_sessions")
        .select("owner_id,phone_number")
        .eq("session_id", sessionId)
        .maybeSingle()
      if (data) await setStatus(sessionId, data.owner_id, data.phone_number, "disconnected")
    }

    return false
  }
}

// ── Utilitários ─────────────────────────────────────────────────────────────
export function getActive(): string[] {
  return Array.from(sessions.keys())
}

export async function getSessionStatus(sessionId: string): Promise<string | null> {
  const pending = await redis.get(`tg:pending:${sessionId}`)
  if (pending) return "pending_code"

  if (sessions.has(sessionId)) return "connected"

  const savedStr = loadStringSession(sessionId)
  if (savedStr) return "disconnected"   // existe em disco mas não na memória

  return null
}

async function setStatus(
  sessionId:   string,
  ownerId:     string,
  phoneNumber: string,
  status:      string
): Promise<void> {
  const { error } = await supabase.from("telegram_sessions").upsert({
    session_id:   sessionId,
    owner_id:     ownerId,
    phone_number: phoneNumber,
    status,
    updated_at:   new Date().toISOString(),
    ...(status === "connected"
      ? { connected_at: new Date().toISOString() }
      : {}),
  }, { onConflict: "session_id" })

  if (error) logger.error({ error, sessionId }, "setStatus Telegram falhou")
}

const maskPhone = (n: string) =>
  n.length <= 6 ? "****" : n.slice(0, 4) + "****" + n.slice(-2)

File: module-telegram-python/src/session_manager.py
"""
session_manager.py

Gerencia sessões Telegram via Telethon (MTProto User API, NÃO bot).
As sessões são serializadas como string e armazenadas no Redis —
igual ao padrão do module-whatsapp que salva credenciais no Redis.

Fluxo (idêntico ao WhatsApp):
  1. POST /session        → startLogin()   → envia código SMS/app
  2. POST /session/confirm → confirmCode() → autentica e persiste sessão
  3. GET  /qr/:sessionId  → getSessionStatus()
  4. Worker chama         → sendMessage()

Chaves Redis usadas:
  tg:session:<sessionId>   → StringSession serializada (sem TTL — persiste)
  tg:pending:<sessionId>   → phone_hash aguardando código (TTL 300s)
  tg:owner:<sessionId>     → owner_id mapeado (TTL 7 dias)
  tg:phone:<sessionId>     → número do chip (TTL 7 dias)
"""
import asyncio
import logging
from typing import Optional

from telethon import TelegramClient
from telethon.sessions import StringSession
from telethon.errors import (
    PhoneCodeInvalidError,
    PhoneCodeExpiredError,
    SessionPasswordNeededError,
    AuthKeyUnregisteredError,
    UserDeactivatedBanError,
)
from telethon.tl.functions.auth import SendCodeRequest
from telethon.tl.types import CodeSettings

from config import settings
from redis_client import get_redis
from supabase_client import get_supabase

logger = logging.getLogger("telegram.session")

# ── Sessões vivas em memória ─────────────────────────────────────────────────
# sessionId → TelegramClient autenticado
_sessions: dict[str, TelegramClient] = {}

# Sessões no meio do login (aguardando código)
# sessionId → {"client": TelegramClient, "phone": str, "phone_hash": str, "owner_id": str}
_pending: dict[str, dict] = {}

TTL_PENDING = 300       # 5 min para digitar o código
TTL_SESSION = 7 * 86400 # 7 dias para owner/phone lookup


# ── Fábrica de cliente ───────────────────────────────────────────────────────
def _make_client(session_str: str = "") -> TelegramClient:
    return TelegramClient(
        StringSession(session_str),
        settings.telegram_api_id,
        settings.telegram_api_hash,
        device_model="ZapFlow",
        system_version="Linux",
        app_version="2.0.0",
        lang_code="pt",
        system_lang_code="pt-BR",
    )


# ── Persistência Redis ───────────────────────────────────────────────────────
async def _save_session(session_id: str, session_str: str) -> None:
    r = get_redis()
    await r.set(f"tg:session:{session_id}", session_str)  # sem TTL — persiste


async def _load_session(session_id: str) -> Optional[str]:
    r = get_redis()
    return await r.get(f"tg:session:{session_id}")


async def _delete_session(session_id: str) -> None:
    r = get_redis()
    await r.delete(
        f"tg:session:{session_id}",
        f"tg:owner:{session_id}",
        f"tg:phone:{session_id}",
        f"tg:pending:{session_id}",
    )


async def _save_meta(session_id: str, owner_id: str, phone: str) -> None:
    r = get_redis()
    pipe = r.pipeline()
    pipe.set(f"tg:owner:{session_id}", owner_id, ex=TTL_SESSION)
    pipe.set(f"tg:phone:{session_id}", phone,    ex=TTL_SESSION)
    await pipe.execute()


async def _load_meta(session_id: str) -> tuple[Optional[str], Optional[str]]:
    r = get_redis()
    owner_id, phone = await asyncio.gather(
        r.get(f"tg:owner:{session_id}"),
        r.get(f"tg:phone:{session_id}"),
    )
    # fallback: Supabase se Redis expirou
    if not owner_id or not phone:
        sb = get_supabase()
        row = (
            sb.table("telegram_sessions")
            .select("owner_id,phone_number")
            .eq("session_id", session_id)
            .maybe_single()
            .execute()
        )
        if row.data:
            owner_id = owner_id or row.data.get("owner_id")
            phone    = phone    or row.data.get("phone_number")
    return owner_id, phone


# ── Supabase — status ────────────────────────────────────────────────────────
async def _set_status(session_id: str, owner_id: str, phone: str, status: str) -> None:
    from datetime import datetime, timezone
    sb = get_supabase()
    now = datetime.now(timezone.utc).isoformat()
    payload: dict = {
        "session_id":   session_id,
        "owner_id":     owner_id,
        "phone_number": phone,
        "status":       status,
        "updated_at":   now,
    }
    if status == "connected":
        payload["connected_at"] = now

    try:
        sb.table("telegram_sessions").upsert(payload, on_conflict="session_id").execute()
    except Exception as exc:
        logger.warning("setStatus falhou session=%s err=%s", session_id, exc)


# ── Passo 1 — iniciar login ──────────────────────────────────────────────────
async def start_login(session_id: str, owner_id: str, phone: str) -> None:
    """
    Envia o código de verificação para o número.
    Se a sessão já está ativa (Redis tem StringSession válida), restaura direto.
    """
    if session_id in _sessions:
        logger.info("Sessão já ativa em memória session=%s", session_id)
        return

    session_str = await _load_session(session_id)
    client = _make_client(session_str or "")

    await client.connect()

    if await client.is_user_authorized():
        # Sessão Redis ainda válida — restaura sem pedir código
        _sessions[session_id] = client
        await _save_meta(session_id, owner_id, phone)
        await _set_status(session_id, owner_id, phone, "connected")
        logger.info("Sessão Telegram restaurada do Redis session=%s phone=%s", session_id, phone)
        return

    # Solicita o código
    result = await client(SendCodeRequest(
        phone_number=phone,
        api_id=settings.telegram_api_id,
        api_hash=settings.telegram_api_hash,
        settings=CodeSettings(),
    ))

    _pending[session_id] = {
        "client":     client,
        "phone":      phone,
        "phone_hash": result.phone_code_hash,
        "owner_id":   owner_id,
    }

    r = get_redis()
    await r.set(f"tg:pending:{session_id}", phone, ex=TTL_PENDING)

    logger.info("Código Telegram enviado session=%s phone=%s", session_id, _mask(phone))


# ── Passo 2 — confirmar código ───────────────────────────────────────────────
async def confirm_code(session_id: str, code: str) -> None:
    """
    Confirma o código recebido no app/SMS e persiste a sessão no Redis.
    Lança exceções descritivas para o caller tratar via HTTP 400.
    """
    pending = _pending.get(session_id)
    if not pending:
        raise ValueError("Nenhum login pendente para esta sessão")

    client: TelegramClient = pending["client"]
    phone: str             = pending["phone"]
    phone_hash: str        = pending["phone_hash"]
    owner_id: str          = pending["owner_id"]

    try:
        await client.sign_in(phone=phone, code=code, phone_code_hash=phone_hash)

    except PhoneCodeInvalidError:
        raise ValueError("Código inválido. Verifique e tente novamente.")
    except PhoneCodeExpiredError:
        raise ValueError("Código expirado. Inicie o login novamente.")
    except SessionPasswordNeededError:
        # 2FA — por enquanto não suportado (decisão do usuário)
        await client.disconnect()
        _pending.pop(session_id, None)
        r = get_redis()
        await r.delete(f"tg:pending:{session_id}")
        raise ValueError(
            "Esta conta tem verificação em dois fatores (2FA). "
            "Desative o 2FA no Telegram ou use um número sem 2FA."
        )

    # Serializa e salva no Redis
    session_str = client.session.save()
    await _save_session(session_id, session_str)
    await _save_meta(session_id, owner_id, phone)

    _sessions[session_id] = client
    _pending.pop(session_id, None)

    r = get_redis()
    await r.delete(f"tg:pending:{session_id}")

    await _set_status(session_id, owner_id, phone, "connected")
    logger.info("Sessão Telegram conectada session=%s phone=%s", session_id, _mask(phone))


# ── Enviar mensagem ──────────────────────────────────────────────────────────
async def send_message(session_id: str, to: str, text: str) -> bool:
    """
    Envia mensagem usando a sessão do chip.
    Restaura automaticamente do Redis se não estiver em memória.
    Retorna True em sucesso, False em falha (worker lança exceção se False).
    """
    client = _sessions.get(session_id)

    if client is None:
        session_str = await _load_session(session_id)
        if not session_str:
            logger.warning("Sessão não encontrada session=%s", session_id)
            return False

        owner_id, phone = await _load_meta(session_id)
        if not owner_id:
            logger.warning("owner_id não encontrado session=%s", session_id)
            return False

        client = _make_client(session_str)
        await client.connect()

        if not await client.is_user_authorized():
            logger.warning("Sessão expirada session=%s", session_id)
            await _set_status(session_id, owner_id, phone or "", "disconnected")
            await _delete_session(session_id)
            return False

        _sessions[session_id] = client
        logger.info("Sessão Telegram restaurada para envio session=%s", session_id)

    try:
        # Telethon aceita número E.164 com ou sem +
        recipient = to if to.startswith("+") else f"+{to}"
        await client.send_message(recipient, text)
        logger.info("✓ Telegram enviado session=%s to=%s", session_id, _mask(to))
        return True

    except (AuthKeyUnregisteredError, UserDeactivatedBanError) as exc:
        logger.error("Sessão revogada session=%s err=%s", session_id, exc)
        _sessions.pop(session_id, None)
        owner_id, phone = await _load_meta(session_id)
        if owner_id:
            await _set_status(session_id, owner_id, phone or "", "disconnected")
        await _delete_session(session_id)
        return False

    except Exception as exc:
        logger.error("✗ Falha Telegram session=%s to=%s err=%s", session_id, _mask(to), exc)
        return False


# ── Utilitários ──────────────────────────────────────────────────────────────
def get_active() -> list[str]:
    return list(_sessions.keys())


async def get_session_status(session_id: str) -> Optional[str]:
    r = get_redis()

    if await r.exists(f"tg:pending:{session_id}"):
        return "pending_code"

    if session_id in _sessions:
        return "connected"

    session_str = await _load_session(session_id)
    if session_str:
        return "disconnected"  # existe no Redis mas não na memória

    return None


async def disconnect_all() -> None:
    """Chamado no shutdown — desconecta todos os clientes graciosamente."""
    for sid, client in list(_sessions.items()):
        try:
            await client.disconnect()
            logger.info("Sessão desconectada no shutdown session=%s", sid)
        except Exception:
            pass
    _sessions.clear()


def _mask(n: str) -> str:
    if not n or len(n) <= 6:
        return "****"
    return n[:4] + "****" + n[-2:]
